manual/tcpechoserver.1
.TH TCPEchoServer 1 "2024 December 03" ""
.SH NAME
tcpechoserver \- Simple TCP echo server
.SH SYNOPSIS
.B tcpechoserver
[port]
.br
.B tcpechoserver
[-h|--help|-v|-version]
.SH COPYRIGHT
TCPEchoServer is Copyright (C) 2024-2025 by Pozsar Zsolt.
.SH DESCRIPTION
This utility is part of ModShell and can be used to test TCP connectivity. The
program is designed for Windows and Unix-like operating systems.
.SH OPTIONS
.TP
.B \-h \-\-help
Write useable parameters.
.TP
.B \-v \-\-version
Write version and build information.
.TP
.B port
Set port number (0-65535).
.SH EXIT STATUS
.TP
.B 0
Normal exit.
.TP
.B 1
Bad command line parameter.
.TP
.B 2
The socket cannot be opened.
.SH HOMEPAGE
.UR http://www.pozsarzs.hu
.UE
.PP
.UR https://pozsarzs.github.io/modshell
.UE
.PP
.UR https://github.com/pozsarzs/modshell
.UE
.SH SEE ALSO
.PD 0
.LP
\fBmbconv\fP(1)
\fBmodshell\fP(1)
\fBnumconv\fP(1)
\fBserialechoserver\fP(1)
\fBserialmbmonitor\fP(1)
\fBudpechoserver\fP(1)
\fBxmodshell\fP(1)
and files in document or /usr/share/doc/modshell/ folder (or equivalent on your system).
.SH AUTHOR
Pozsar Zsolt
.MT pozsarzs@gmail.com
.ME


document/changelog
modshell (0.1-beta1) unstable; urgency=low

  * First user test release will be with the following changes:
  * Elimination of confusion between the terms 'register number' and 'data
    address';
  * in interpreter mode, passing parameters from the OS command line to the
    script and returning the script's exit value to the OS;
  * in interpreter mode, command verbosity level setting with a predefined
    variable;
  * ANSI escape sequences support (only DOS and *nix like systems);
  * command `color` -> `set color`, sets all default colors (CLI and TUI);  
  * command `echo` -> `echometh`, parameters: off|on|hex|swap -> off|an|hex|swap;
  * commands `sercons`, `tcpcons`, `udpcons`: character-to-character or string
    sending, with alphanumerical or hexadecimal input, with or without
    alphanumerical or hexhexadecimal echo;  
  * commands `serwrite`, `tcpwrite`, `udpwrite`: alphanumerical or hexadecimal
    input, with or without alphanumerical or hexadecimal echo;
  * DCON protocol support;  
  * keywords (coil, hreg, asc, hex, etc.) should also be specified from
    variables;
  * main menu for all consoles in GUI version;  
  * Modbus register number/address converter utility (in your own scripting
    language);
  * Number converter utility (in your own scripting language);
  * modified source code structure of XModshell;  
  * new command `chkdevlock`/`rmdevlock` (only *nix systems);  
  * new command `datatype`;  
  * new command `exist`;  
  * new command `inputmeth`;  
  * new command `macro`;
  * new command `mbconv`;  
  * new command `mbmon`;  
  * new command `printcolor`(only CLI and TUI);  
  * new command `runmeth`;  
  * new command `sendmeth`;  
  * new command `tcpcons`, `tcpread`, `tcpwrite`;  
  * new command `udpcons`, `udpread`, `udpwrite`;  
  * new commands `ioread` and `iowrite`;
  * new menu items in the main menu for quick execution of Modbus R/W commands;  
  * new menu items for show and edit all register's value in a big table;  
  * new serial baudrates: 150, 300, 600 baud;  
  * serial echo server utility for testing connectivity;  
  * serial ModBus traffic monitor utility;
  * support for variable and constant arrays;  
  * new predefined constants;  
  * syntax highlighter file for applications using GTKSourceView (for example:
    Builder, Geany, Gedit, Mousepad, Pluma, Scribes), for Notepad++ (Windows
    only) and Vim/Neovim;
  * TCP and UPD echo server utilities for testing connectivity.
 
 -- Pozsár Zsolt <pozsarzs@gmail.com> Wed, 01 Jan 2025 00:00:00

modshell (0.1-alpha3) unstable; urgency=low

  * Next developer test release.
  * New source code directory structure;
  * new commands;
  * new predefined constants;
  * bug fixes;
  * GUI (except DOS);
  * script syntax highlighting file for MCEdit and Nano.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Sun, 20 Oct 2024 09:37:44
 
modshell (0.1-alpha2) unstable; urgency=low

  * Next developer test release.
  * Bug fixes;
  * Modbus/ASCII and Modbus/RTU communication;
  * handling of constants;
  * three predefined constants;
  * 28 new command (total: 94);
  * script syntax highlighting file for Micro.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Mon, 12 Feb 2024 00:00:00

modshell (0.1-alpha1) unstable; urgency=low

  * Initial release.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Fri, 16 Jan 2024 21:12:00


source/tcpechoserver/tcpechoserver.pas
{ +--------------------------------------------------------------------------+ }
{ | ModShell v0.1 * Command-driven scriptable Modbus utility                 | }
{ | Copyright (C) 2023-2025 Pozsar Zsolt <pozsarzs@gmail.com>                | }
{ | tcpechoserver.pas                                                        | }
{ | TCP echo server utility                                                  | }
{ +--------------------------------------------------------------------------+ }
{ 
  This program is free software: you can redistribute it and/or modify it
  under the terms of the European Union Public License 1.2 version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.
}

{$IFDEF GO32V2} {$ERROR "Cannot compile on this system." } {$ENDIF}
{$IFDEF WINDOWS} {$APPTYPE CONSOLE} {$ENDIF}
{$MODE OBJFPC}{$H+}{$MACRO ON}

program tcpechoserver;
uses
  BlckSock,
  SysUtils,
  crt,
  ucommon,
  utranslt;
var
  Socket1, Socket2: TTCPBlockSocket;
  appmode: byte;
  b: byte;
  c: char;
  lang: string;
  port: string = '';
  s: string;
const
  // COMMAND LINE ARGUMENTS
  CMDLINEPARAMS: array[0..1, 0..2] of string =
  (
    ('-h','--help','show help'),
    ('-v','--version','show version and build information')
  );
  // OTHERS
  PRGCOPYRIGHT = '(C) 2024 Pozsar Zsolt <http://www.pozsarzs.hu>';
  PRGNAME = 'TCPEchoServer';
  PRGVERSION = '0.1';

{$DEFINE BASENAME := lowercase(PRGNAME)}

{$R *.res}

resourcestring
  // messages
  MSG02 = 'TCP Echo Server utility';
  MSG03 = 'port: ';
  MSG04 = 'Press [ESC] to exit.';
  MSG05 = 'Server running...';
  MSG06 = 'message: ';
  MSG07 = 'Server stopped.';
  MSG94 = 'Build date:  ';
  MSG95 = 'Builder:     ';
  MSG96 = 'FPC version: ';
  MSG97 = 'Target OS:   ';
  MSG98 = 'Target CPU:  ';
  // error messages
  ERR01 = 'ERROR: ';
  ERR02 = 'The port value must be between 0-65535.';

// SHOW USEABLE ARGUMENTS
procedure help(mode: boolean);
var
  b: byte;
begin
  if mode
    then writeln('There are one or more bad argument in command line.') else
  begin
    writeln('Usage: ' + BASENAME + ' [port]');
    writeln('       ' + BASENAME + ' [argument]');
    writeln;
    writeln('arguments:');
    for b := 0 to 1 do
    begin
      write('  ',CMDLINEPARAMS[b, 0]);
      gotoxy(8, wherey); write(CMDLINEPARAMS[b, 1]);
      gotoxy(30, wherey); writeln(CMDLINEPARAMS[b, 2]);
    end;
    writeln;
  end;
  quit(0, false, '');
end;

{$I version.pas}

begin
  // detect language
  lang := getlang;
  translatemessages(LANG, BASENAME, '.mo');
  // command line arguments
  appmode := 0;
  { appmode #0: normal run
    appmode #1: show useable arguments
    appmode #2: show version and build information }
  if paramcount > 0 then
  begin
    for b := 0 to 1 do
    begin
      if paramstr(1) = CMDLINEPARAMS[b, 0] then appmode := b + 1;
      if paramstr(1) = CMDLINEPARAMS[b, 1] then appmode := b + 1;
    end;
    case appmode of
      1: help(false);
      2: version(true);
    end;
  end;
  writeln(PRGNAME + ' v' + PRGVERSION + ' * ' + MSG02);
  writeln(PRGCOPYRIGHT);
  writeln;
  writeln(MSG04);
  // port number
  if paramcount = 0 then
  begin
    write(MSG03 + '?');
    repeat
      c := readkey;
      if (ord(c) >= 48) and (ord(c) <= 57) and
         (strtoint(port + c) <= 65535) then port := port + c;
      if (c = #8) and (length(port) > 0) then delete(port,length(port), 1);
      gotoxy(1, wherey); clreol;
      write(MSG03 + port);
    until ((length(port) > 0) and (c = #13)) or (c = #27);
    if c = #27 then quit(0, false, #13);
  end else port := paramstr(1);
  writeln;
  // check port number
  if not ((strtointdef(port, -1) >= 0) and (strtointdef(port, -1) <= 65535))
    then quit(1, false, ERR01 + ERR02);
  // PRIMARY MISSION
  gotoxy(1, wherey); clreol;
  writeln(MSG03 + port);
  Socket1 := TTCPBlockSocket.Create;
  Socket1.Bind('0.0.0.0', port);
  if Socket1.LastError <> 0 then quit(2, false, ERR01 + Socket1.LastErrorDesc);
  Socket1.Listen;
  writeln(MSG05);
  repeat
    if Socket1.CanRead(100) then
    begin
      Socket2 := TTCPBlockSocket.Create;
      Socket2.Socket := Socket1.Accept;
      s := Socket2.RecvPacket(1000);
      if length(s) > 0 then
        if Socket1.LastError = 0 then
        begin
          Socket2.SendString(s);
          writeln(MSG06 + '"' + s + '"');
        end;
      Socket2.CloseSocket;
      Socket2.Free;
      delay(100);
    end;
    if keypressed then c := readkey;
  until c = #27;
  Socket1.CloseSocket;
  Socket1.Free;
  writeln(MSG07);
  quit(0, false, '');
end.
